internal/app/app.go

```diff
@@ -0,0 +1,159 @@
+// Package app exports a D-Bus object to apply proxy settings.
+package app
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/godbus/dbus/v5/introspect"
+	log "github.com/sirupsen/logrus"
+	"github.com/ubuntu/decorate"
+	"github.com/ubuntu/ubuntu-proxy-manager/internal/proxy"
+)
+
+const dbusObjectPath = "/com/ubuntu/ProxyManager"
+const dbusInterface = "com.ubuntu.ProxyManager"
+
+// DryRun is a context key to indicate that we should not apply the proxy
+// settings. It is intended to be used in tests only.
+type DryRun struct{}
+
+type busObject struct {
+	ctx    context.Context
+	cancel context.CancelCauseFunc
+
+	mu sync.Mutex
+}
+
+type options struct {
+	timeout time.Duration
+	initCtx context.Context
+}
+type option func(*options)
+
+// App is the main application object.
+type App struct {
+	conn    *dbus.Conn
+	obj     *busObject
+	timeout time.Duration
+}
+
+// Apply is a function called via D-Bus to apply the system proxy settings.
+func (b *busObject) Apply(http, https, ftp, socks, no, mode string) *dbus.Error {
+	// Methods calls spin up separate goroutines, so ensure we don't run them in parallel
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	log.Infof("Called Apply(%s, %s, %s, %s, %s, %s)", http, https, ftp, socks, no, mode)
+
+	p, err := proxy.New(b.ctx, http, https, ftp, socks, no, mode)
+	// We need to cancel the context in a deferred function to get the final
+	// state of the error variable.
+	defer func() { b.cancel(err) }()
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	if b.ctx.Value(DryRun{}) == true {
+		return nil
+	}
+
+	err = p.Apply()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// New creates a new App object.
+func New(opts ...option) (a *App, err error) {
+	defer decorate.OnError(&err, "cannot initialize application")
+
+	// defaults
+	args := options{
+		timeout: 1 * time.Second,
+		initCtx: context.Background(),
+	}
+	// applied options
+	for _, o := range opts {
+		o(&args)
+	}
+
+	// Don’t call dbus.SystemBus which caches globally system dbus (issues in tests)
+	conn, err := dbus.SystemBusPrivate()
+	if err != nil {
+		return nil, err
+	}
+	if err = conn.Auth(nil); err != nil {
+		_ = conn.Close()
```

Comment:
If we are ignoring the error (I'm not sure why the linter is not complaining about it), wdyt about something like `defer conn.Close()`? I think it would look more like the way we handle these situations in our other projects.